### PR TITLE
feat: add nodegroup resource limits

### LINKED ISSUE #4954: Support for Queue-level Quotas per Node Group to enable logical resource overflow.
### What is the problem you're trying to solve

We are managing multiple queues (1, 2, 3) across two node groups (A and B). 
We need to enforce a logical resource limit (x CPUs/Memory) for Queue 1 specifically on Node Group A. 
Once Queue 1 exceeds x CPUs/Memory on Node Group A, any additional workloads should automatically overflow to Node Group B.

Currently, Volcano only triggers overflow when Node Group A is physically exhausted. There is no mechanism to define a per-queue capacity or soft-constraint specifically mapped to a Node Group, making it impossible to redirect traffic based on a logical threshold before the physical nodes are full.

### Describe the solution you'd like

Support set capacity at queue-level for each nodegroup

### Additional context

_No response_

### Documentation Updates

- [x] This feature requires design or user documentation changes.
- [x] If documentation changes are required, I will ensure the relevant documents are updated and published to the Volcano official website (https://volcano.sh) via the [volcano-sh/website](https://github.com/volcano-sh/website) repository.

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Summary

- Add `volcano.sh/nodegroup-resource-limits` queue annotation to cap queue usage per nodegroup.
- Track allocated resources by `queue + nodegroup` in the nodegroup plugin and reject nodes when future usage would exceed the configured limit.
- Add unit tests for no annotation, under-limit scheduling, fallback after preferred nodegroup reaches its logical limit, invalid annotation handling, and allocation event accounting.
- Document the new annotation in the nodegroup plugin user guide.

Closes #4954

## Test Plan

- [x] `go test ./pkg/scheduler/plugins/nodegroup`

### CODE REVIEW

File: pkg/scheduler/plugins/nodegroup/nodegroup_test.go [hunk 1]
```diff
@@ -382,3 +382,223 @@ func TestNodeGroup(t *testing.T) {
...
+
+			var targetTask *api.TaskInfo
+			for _, job := range ssn.Jobs {
+				for _, task := range job.Tasks {
+					if task.Name == test.targetTaskName {
+						targetTask = task
+						break
+					}
+				}
+			}
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup_test.go:568]: Fixed in the latest push with a labeled break once the target task is found.

File: pkg/scheduler/plugins/nodegroup/nodegroup.go [hunk 2]
```diff
@@ -300,6 +413,9 @@ func (np *nodeGroupPlugin) OnSessionOpen(ssn *framework.Session) {
 			return fmt.Errorf("job %s not found in session", task.Job)
 		}
 		attr := np.queueAttrs[job.Queue]
+		if attr.resourceLimitErr != nil {
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup.go]: Fixed in the latest push by guarding attr before checking resourceLimitErr.

File: pkg/scheduler/plugins/nodegroup/nodegroup_test.go [hunk 3]
```diff
@@ -382,3 +382,223 @@ func TestNodeGroup(t *testing.T) {
...
+			if targetTask == nil {
+				t.Fatalf("target task %s not found", test.targetTaskName)
+			}
+
+			for _, node := range ssn.Nodes {
+				err := ssn.PredicateFn(targetTask, node)
+				code := api.Success
+				if err != nil {
+					code = err.(*api.FitError).Status[0].Code
+				}
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup_test.go]: Fixed in the latest push by using errors.As and failing the test with the unexpected error type.

File: pkg/scheduler/plugins/nodegroup/nodegroup.go [hunk 4]
```diff
@@ -300,6 +413,9 @@ func (np *nodeGroupPlugin) OnSessionOpen(ssn *framework.Session) {
...
+			return newFitErr(task, node, errNodeGroupResourceLimitInvalid)
+		}
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup.go:457]: Fixed in the latest push by making the resourceLimitErr check nil-safe before reading it.

File: pkg/scheduler/plugins/nodegroup/nodegroup_test.go [hunk 5]
```diff
@@ -382,3 +383,233 @@ func TestNodeGroup(t *testing.T) {
...
+func fitErrorCode(t *testing.T, err error) int {
+	t.Helper()
+	if err == nil {
+		return api.Success
+	}
+	var fitErr *api.FitError
+	if !errors.As(err, &fitErr) {
+		t.Fatalf("expected *api.FitError, got %T: %v", err, err)
+	}
+	return fitErr.Status[0].Code
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup_test.go:596]: Fixed in the latest push by checking that FitError.Status is non-empty before reading the first status code.

File: pkg/scheduler/plugins/nodegroup/nodegroup.go [hunk 6]
```diff
@@ -257,14 +289,95 @@ func (np *nodeGroupPlugin) updateAffinityFromAncestor(attr *queueAttr) {
...
+	}
+
+	allocated := attr.nodeGroupAllocated[group]
+	if allocated == nil {
+		allocated = api.EmptyResource()
+	}
+	futureUsed := allocated.Clone().Add(task.Resreq)
+	if ok, resourceNames := futureUsed.LessEqualWithDimensionAndResourcesName(limit, task.Resreq); !ok {
+		return fmt.Errorf("nodegroup %s resource limit exceeded, insufficient resources: %v", group, resourceNames)
+	}
```
[the remark this answers is not in the document]
Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup.go:375]: Fixed in the latest push by preserving the resource keys explicitly configured for each nodegroup and comparing only those dimensions. Added CPU-only limit tests for omitted memory and CPU enforcement.

File: pkg/scheduler/plugins/nodegroup/nodegroup.go [hunk 7]
```diff
@@ -34,6 +36,8 @@ const (
 	NodeGroupNameKey = "volcano.sh/nodegroup-name"
 	rootQueueID      = "root"
 
+	nodeGroupResourceLimitsAnnotationKey = "volcano.sh/nodegroup-resource-limits"
```
Comment (hajnalmt) [inline pkg/scheduler/plugins/nodegroup/nodegroup.go]: Since both of these are user-facing `volcano.sh/*` keys, can we move them to the official scheduling API constants under `staging/src/volcano.sh/apis/pkg/apis/scheduling/v1beta1/labels.go` and reference them from here? `NodeGroupNameKey` is already used as an external node label key, and `volcano.sh/nodegroup-resource-limits` is now a documented Queue annotation, so keeping the literals private to the plugin makes the public API surface harder to discover and easier to duplicate later.

**Review by hajnalmt — COMMENTED** [review]
/area scheduling

Thanks for the PR, cool feature, this is a nice addition to the nodegroup plugin. The limit accounting path is now much clearer, the edge cases from the previous bot comments look covered, and the docs/tests make the new Queue annotation behavior easy to follow. I left one API-surface comment about where the public `volcano.sh/*` keys should live before this gets finalized.

Great work!

File: pkg/scheduler/plugins/nodegroup/nodegroup.go (hunk 7 above)
Comment (JesseStutler) [reply pkg/scheduler/plugins/nodegroup/nodegroup.go]: Yes export this as a const in the `staging/src/volcano.sh/apis` just as mate said will be better

Comment (ruanwenjun) [reply pkg/scheduler/plugins/nodegroup/nodegroup.go]: Moved both user-facing keys to scheduling v1beta1 API constants in labels.go and updated the nodegroup plugin/tests to reference schedulingv1.NodeGroupNameKey and schedulingv1.NodeGroupResourceLimitsAnnotationKey.

**Comment by JesseStutler** [comment]
@ruanwenjun Could you also update the nodegroup plugin doc on the website? Thanks

**Review by hajnalmt — APPROVED** [review]
/lgtm

Super thanks! Please don't forget to update the website too!
